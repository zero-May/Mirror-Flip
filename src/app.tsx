import "/styles/main.css";
import { useState } from "react";
import { addElementAtPoint, requestExport } from "@canva/design";
import { upload } from "@canva/asset";
import { Button, Rows, Title, Text, Box, Slider, FileInput } from "@canva/app-ui-kit";

export const App = () => {
  const [image, setImage] = useState(null);
  const [opacityValue, setOpacityValue] = useState(1);

  const canvas = document.createElement("canvas");
  const ctx = canvas.getContext("2d");
  const img = new Image();

  // 上传图片
  const onFileChange = (event) => {
    const file = event[0];
    if (file) {
      const reader = new FileReader();
      reader.onload = (e) => {
        setImage(e.target.result);
      };
      reader.readAsDataURL(file);
    }
  };

  // 翻转图片
  const flipImage = (direction) => {
    img.onload = () => {
      canvas.width = img.width;
      canvas.height = img.height;
      if (direction === "horizontal") {
        ctx.translate(img.width, 0);
        ctx.scale(-1, 1);
      } else {
        ctx.translate(0, img.height);
        ctx.scale(1, -1);
      }
      ctx.globalAlpha = opacityValue; // 设置透明度
      ctx.drawImage(img, 0, 0);
      setImage(canvas.toDataURL());
    };
    img.src = image;
  };

  // 改变透明度
  const changeOpacity = (value) => {
    setOpacityValue(value);
  };

  // 添加到画板
  async function addToDesign() {
    img.onload = () => {
      canvas.width = img.width;
      canvas.height = img.height;
      ctx.globalAlpha = opacityValue; // 设置透明度
      ctx.drawImage(img, 0, 0);
      setImage(canvas.toDataURL());
    };

    const result = await upload({
      type: "image",
      mimeType: "image/png",
      url: image,
      thumbnailUrl: image,
      aiDisclosure: "none",
    });

    // 添加到页面
    await addElementAtPoint({
      type: "image",
      ref: result.ref,
      altText: {
        text: "Example image",
        decorative: false,
      },
    });
  }

  // 导出图片
  async function exportImage() {
    const result = await requestExport({
      acceptedFileTypes: ["jpg", "png"],
    });
    console.log(result); // => { status: "complete", title: "My design", exportBlobs: [{ url: "https://example.com/image.png" }] }
  }

  return (
    <Rows spacing="2u">
      <Title>镜像翻转迷你程序</Title>
      <FileInput accept={["image/*"]} onDropAcceptedFiles={onFileChange} />
      {image && (
        <>
          <Title>预览图</Title>
          <img src={image} alt="Uploaded" style={{ opacity: opacityValue }} />
          <Rows spacing="2u">
            <Button
              alignment="center"
              ariaLabel="Label text"
              onClick={() => flipImage("horizontal")}
              variant="primary"
            >
              水平翻转
            </Button>
            <Button
              alignment="center"
              ariaLabel="Label text"
              onClick={() => flipImage("vertical")}
              variant="primary"
            >
              垂直翻转
            </Button>
          </Rows>
          <Rows spacing="2u">
            <Text>调整透明度:</Text>
            <Box paddingStart="2u">
              <Slider
                defaultValue={opacityValue}
                max={1}
                min={0}
                step={0.1}
                onChange={changeOpacity}
              />
            </Box>
          </Rows>
          <Button
            alignment="center"
            ariaLabel="Label text"
            onClick={addToDesign}
            variant="primary"
          >
            添加到页面
          </Button>
          <Button
            alignment="center"
            ariaLabel="Label text"
            onClick={exportImage}
            variant="primary"
          >
            导出图片
          </Button>
        </>
      )}
    </Rows>
  );
};
