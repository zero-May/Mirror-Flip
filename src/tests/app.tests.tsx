/* eslint-disable formatjs/no-literal-string-in-jsx */
import React from "react";
import type { RenderResult } from "@testing-library/react";
import { fireEvent, render } from "@testing-library/react";
import { TestAppI18nProvider } from "@canva/app-i18n-kit";
import { TestAppUiProvider } from "@canva/app-ui-kit";
import { requestOpenExternalUrl } from "@canva/platform";
import { useAddElement } from "utils/use_add_element";
import { App, DOCS_URL } from "../app";

function renderInTestProvider(node: React.ReactNode): RenderResult {
  return render(
    // In a test environment, you should wrap your apps in `TestAppI18nProvider` and `TestAppUiProvider`, rather than `AppI18nProvider` and `AppUiProvider`
    <TestAppI18nProvider>
      <TestAppUiProvider>{node}</TestAppUiProvider>,
    </TestAppI18nProvider>,
  );
}

jest.mock("utils/use_add_element");

// This test demonstrates how to test code that uses functions from the Canva Apps SDK
// For more information on testing with the Canva Apps SDK, see https://www.canva.dev/docs/apps/testing/
describe("Hello World Tests", () => {
  // Mocking the useAddElement hook
  const mockAddElement = jest.fn();
  const mockAddUseElement = jest.mocked(useAddElement);
  const mockRequestOpenExternalUrl = jest.mocked(requestOpenExternalUrl);

  beforeEach(() => {
    jest.resetAllMocks();
    mockAddUseElement.mockReturnValue(mockAddElement);
    mockRequestOpenExternalUrl.mockResolvedValue({ status: "completed" });
  });

  // this test uses a mock in place of the useAddElement hook
  it("should add a text element when the button is clicked", () => {
    // assert that the mocks are in the expected clean state
    expect(mockAddUseElement).not.toHaveBeenCalled();
    expect(mockAddElement).not.toHaveBeenCalled();

    const result = renderInTestProvider(<App />);

    // the hook should have been called in the render process but not the callback
    expect(mockAddUseElement).toHaveBeenCalled();
    expect(mockAddElement).not.toHaveBeenCalled();

    // get a reference to the do something cool button element
    const doSomethingCoolBtn = result.getByRole("button", {
      name: "Do something cool",
    });

    // programmatically simulate clicking the button
    fireEvent.click(doSomethingCoolBtn);

    // we expect that addElement has been called by the button's click handler
    expect(mockAddElement).toHaveBeenCalled();
  });

  // this test uses a mock in place of the @canva/platform requestOpenExternalUrl function
  it("should call `requestOpenExternalUrl` when the button is clicked", () => {
    expect(mockRequestOpenExternalUrl).not.toHaveBeenCalled();

    const result = renderInTestProvider(<App />);

    // get a reference to the Apps SDK button by name
    const sdkButton = result.getByRole("button", {
      name: "Open Canva Apps SDK docs",
    });

    expect(mockRequestOpenExternalUrl).not.toHaveBeenCalled();
    fireEvent.click(sdkButton);
    expect(mockRequestOpenExternalUrl).toHaveBeenCalled();

    // assert that the requestOpenExternalUrl function was called with the expected arguments
    expect(mockRequestOpenExternalUrl.mock.calls[0][0]).toEqual({
      url: DOCS_URL,
    });
  });

  // this test demonstrates the use of a snapshot test
  it("should have a consistent snapshot", () => {
    const result = renderInTestProvider(<App />);
    expect(result.container).toMatchSnapshot();
  });
});
