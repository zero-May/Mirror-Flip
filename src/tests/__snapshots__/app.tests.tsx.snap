// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hello World Tests should have a consistent snapshot 1`] = `
<div>
  <div
    class="scrollContainer"
  >
    <div
      class="jv_R6g"
      style="--NZu1Zw: 16px;"
    >
      <p
        class="D6xAvw E44aNA rU8RQQ"
      >
        To make changes to this app, edit the 
        <code>
          src/app.tsx
        </code>
         file, then close and reopen the app in the editor to preview the changes.
      </p>
      <button
        class="_5KtATA LQzFZw xwTbNA _8ERLTg rV61VQ LQzFZw xwTbNA _4_iekA _3oHQrw j34Zww aqbYPg _3FvZZg"
        type="button"
      >
        <span
          class="khPe7Q"
        >
          Do something cool
        </span>
      </button>
      <button
        class="_5KtATA LQzFZw xwTbNA _8ERLTg Z3nT2A LQzFZw xwTbNA _4_iekA j34Zww aqbYPg _3FvZZg"
        type="button"
      >
        <span
          class="khPe7Q"
        >
          Open Canva Apps SDK docs
        </span>
      </button>
    </div>
  </div>
  ,
</div>
`;
